From af283471cc8f3b8fb4cda139e28e0e8b9d5fa021 Mon Sep 17 00:00:00 2001
From: Meghana H <Meghana.H@arm.com>
Date: Fri, 17 Apr 2026 14:19:47 +0000
Subject: [PATCH] redfish: Add FaultLog attachment download support

Add support in bmcweb to expose FaultLog attachments
through Redfish. Include AdditionalDataURI and
AdditionalDataSizeBytes for FaultLog entries and add
a download route for attachments.

Allow FaultLog entries to use the existing dump attachment
download path.This enables Redfish clients to download
the raw CPER binary associated with a FaultLog entry
while keeping existing behavior unchanged for other
dump entry types.

Tested :

'''

Generate a cper using libcper:

./libcper/build$ ./cper-generate --out test-cper.bin --sections generic

Copy the cper bin into BMC:

scp test-cper.bin   root@{BMC-IP}:/var/cper/

Trigger CreateDump:

busctl call xyz.openbmc_project.Dump.Manager  /xyz/openbmc_project/dump/faultlog  xyz.openbmc_project.Dump.Create CreateDump  "a{sv}" 2    "Type" "s" "CPER"    "Primary
LogId" "s" "/var/cper/test-cper.bin"

curl -k -u root:0penBmc -X GET https://{BMC_IP}/redfish/v1/Managers/bmc/LogServices/FaultLog/Entries/3

{
  "@odata.id": "/redfish/v1/Managers/bmc/LogServices/FaultLog/Entries/3",
  "@odata.type": "#LogEntry.v1_11_0.LogEntry",
  "AdditionalDataSizeBytes": 792,
  "AdditionalDataURI": "/redfish/v1/Managers/bmc/LogServices/FaultLog/Entries/3/attachment",
  "Created": "2026-04-17T07:07:24.775620+00:00",
  "EntryType": "Event",
  "Id": "3",
  "Name": "FaultLog Dump Entry"
}

Download attachment:

wget --no-check-certificate --user=root --password=0penBmc \
  https://{BMC-IP}/redfish/v1/Managers/bmc/LogServices/FaultLog/Entries/1/attachment

'''

Upstream-Status: Submitted [https://gerrit.openbmc.org/c/openbmc/bmcweb/+/91085]

Signed-off-by: Meghana H <Meghana.H@arm.com>
Change-Id: I086de6fcf79b6c8ed92e7668ead61e4215678cb8
---
 .../include/utils/log_services_utils.hpp      |  5 ++--
 redfish-core/lib/log_services.hpp             | 25 ++++++++++++++++++-
 redfish-core/src/redfish.cpp                  |  1 +
 3 files changed, 28 insertions(+), 3 deletions(-)

diff --git a/redfish-core/include/utils/log_services_utils.hpp b/redfish-core/include/utils/log_services_utils.hpp
index 0a7ca87c..a421861b 100644
--- a/redfish-core/include/utils/log_services_utils.hpp
+++ b/redfish-core/include/utils/log_services_utils.hpp
@@ -81,7 +81,8 @@ inline void downloadEntryCallback(
     }
 
     // Make sure we know how to process the retrieved entry attachment
-    if ((downloadEntryType != "BMC") && (downloadEntryType != "System"))
+    if ((downloadEntryType != "BMC") && (downloadEntryType != "System") &&
+        (downloadEntryType != "FaultLog"))
     {
         BMCWEB_LOG_ERROR("downloadEntryCallback() invalid entry type: {}",
                          downloadEntryType);
@@ -101,7 +102,7 @@ inline void downloadEntryCallback(
         close(fd);
         return;
     }
-    if (downloadEntryType == "System")
+    if (downloadEntryType == "System" || downloadEntryType == "FaultLog")
     {
         if (!asyncResp->res.openFd(fd))
         {
diff --git a/redfish-core/lib/log_services.hpp b/redfish-core/lib/log_services.hpp
index 66ddd087..19e28805 100644
--- a/redfish-core/lib/log_services.hpp
+++ b/redfish-core/lib/log_services.hpp
@@ -347,6 +347,12 @@ inline void getDumpEntryCollection(
                         "{}/{}/attachment", entriesPath, entryID);
                     thisEntry["AdditionalDataSizeBytes"] = size;
                 }
+                else if (dumpType == "FaultLog")
+                {
+                    thisEntry["AdditionalDataURI"] = boost::urls::format(
+                        "{}/{}/attachment", entriesPath, entryID);
+                    thisEntry["AdditionalDataSizeBytes"] = size;
+                }
                 entriesArray.emplace_back(std::move(thisEntry));
             }
             asyncResp->res.jsonValue["Members@odata.count"] =
@@ -447,6 +453,13 @@ inline void getDumpEntryById(
                                             entryID);
                     asyncResp->res.jsonValue["AdditionalDataSizeBytes"] = size;
                 }
+                else if (dumpType == "FaultLog")
+                {
+                    asyncResp->res.jsonValue["AdditionalDataURI"] =
+                        boost::urls::format("{}/{}/attachment", entriesPath,
+                                            entryID);
+                    asyncResp->res.jsonValue["AdditionalDataSizeBytes"] = size;
+                }
             }
             if (!foundDumpEntry)
             {
@@ -491,7 +504,7 @@ inline void downloadDumpEntry(
     const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
     const std::string& entryID, const std::string& dumpType)
 {
-    if (dumpType != "BMC" && dumpType != "System")
+    if (dumpType != "BMC" && dumpType != "System" && dumpType != "FaultLog")
     {
         BMCWEB_LOG_WARNING("Can't find Dump Entry {}", entryID);
         messages::resourceNotFound(asyncResp->res, dumpType + " dump", entryID);
@@ -1587,6 +1600,16 @@ inline void requestRoutesFaultLogDumpEntry(App& app)
             handleLogServicesDumpEntryDelete, std::ref(app), "FaultLog"));
 }
 
+inline void requestRoutesFaultLogDumpEntryDownload(App& app)
+{
+    BMCWEB_ROUTE(
+        app,
+        "/redfish/v1/Managers/<str>/LogServices/FaultLog/Entries/<str>/attachment/")
+        .privileges(redfish::privileges::getLogEntry)
+        .methods(boost::beast::http::verb::get)(std::bind_front(
+            handleLogServicesDumpEntryDownloadGet, std::ref(app), "FaultLog"));
+}
+
 inline void requestRoutesFaultLogDumpClear(App& app)
 {
     BMCWEB_ROUTE(
diff --git a/redfish-core/src/redfish.cpp b/redfish-core/src/redfish.cpp
index a9601000..ea606250 100644
--- a/redfish-core/src/redfish.cpp
+++ b/redfish-core/src/redfish.cpp
@@ -171,6 +171,7 @@ RedfishService::RedfishService(App& app)
         requestRoutesFaultLogDumpService(app);
         requestRoutesFaultLogDumpEntryCollection(app);
         requestRoutesFaultLogDumpEntry(app);
+        requestRoutesFaultLogDumpEntryDownload(app);
         requestRoutesFaultLogDumpClear(app);
     }
 
-- 
2.53.0-Meta

